### DISCUSSION TITLE
col-span-16 with dynamically generated cols

### DISCUSSION THREAD

**Discussion by alexgenaud** [tailwindlabs/tailwindcss#11715] [Help] [answered]
I dynamically build a 16 column-wide grid entirely with React TSX components. The 16 grid layout CSS does not seem to compile unless literal examples of 16 cells in a 16 column grid are discovered (by Tailwind, Vite, TSC, some pre-processor) in some TSX or HTML-like file.

https://github.com/alexgenaud/north/blob/master/src/components/TailwindHack.tsx

A dummy/hack component which includes a row of 16 columns fixes the issue, presumably by hinting to Tailwind that 16 columns do (or will) in fact exist.

Is there a more elegant way?

Working Vite/Tailwind page:
https://alexgenaud.github.io/north/

```
    <div className="grid grid-cols-16 gap-1">
      <div className="col-span-1">1</div>
      <div className="col-span-2">2</div>
      ...
      <div className="col-span-12">12</div>
      ...
      <div className="col-span-16">16</div>
    </div>
```

# Solution (within tailwind.config.js):

```
  safelist: [
    {
      pattern: /^col-span-([1-9]|1[0-6])$/
    },
  ],
  theme: {
    extend: {
      gridTemplateColumns: {
        16: "repeat(16, minmax(0, 1fr))",
      },
      gridColumn: {
        'span-13': 'span 13 / span 13',
        'span-14': 'span 14 / span 14',
        'span-15': 'span 15 / span 15',
        'span-16': 'span 16 / span 16',
      },
    },
  },
```

**Comment by wongjn** [comment]
As per [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as _complete unbroken strings_ in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:
>
> **Don’t construct class names dynamically**
>
> ```vue
> <div class="text-{{ error ? 'red' : 'green' }}-600"></div>
> ```
> In the example above, the strings `text-red-600` and `text-green-600` do not exist, so Tailwind will not generate those classes.
> Instead, make sure any class names you’re using exist in full:
>
> **Always use complete class names**
>
> ```vue
> <div class="{{ error ? 'text-red-600' : 'text-green-600' }}"></div>
> ```

You could consider:
- Using [`safelist`](https://tailwindcss.com/docs/content-configuration#safelisting-classes) correctly.
- Using the `style` attribute to apply the appropriate `grid-column` value.

**Comment by alexgenaud** [comment]
Thanks @wongjn , perhaps I am using safelist incorrectly, but it's not working for me.

Working example with the Hack (described in the op):

<img width="959" alt="0-working-hack" src="https://github.com/tailwindlabs/tailwindcss/assets/32287/e0377e50-4e96-4105-900a-e4ce6473f856">

However, if I remove the dummy/hack TSX and update safelist with explicit patterns (probably incorrectly again):

```
export default {
  content: ["./src/**/*.{html,js,jsx,tsx,ts}"],
  safelist: [
    {
      pattern: /col-span-1/,
      pattern: /col-span-2/,
      pattern: /col-span-3/,
      pattern: /col-span-4/,
      pattern: /col-span-5/,
      pattern: /col-span-6/,
      pattern: /col-span-7/,
      pattern: /col-span-8/,
      pattern: /col-span-9/,
      pattern: /col-span-10/,
      pattern: /col-span-11/,
      pattern: /col-span-12/,
      pattern: /col-span-13/,
      pattern: /col-span-14/,
      pattern: /col-span-15/,
      pattern: /col-span-16/,
    },
  ],
  theme: {
    extend: {
      gridTemplateColumns: {
        // Simple 16 column grid
        16: "repeat(16, minmax(0, 1fr))",
      },
    },
  },
  plugins: [],
};
```

Does not work. Below should have a few col-span-8 examples, but do not:

<img width="537" alt="1 safelist-only-fail" src="https://github.com/tailwindlabs/tailwindcss/assets/32287/c07197c4-3cb7-4016-ae32-2f1398cc8bda">

**Reply by alexgenaud** [reply]
Earlier, my `safelist` contained the patterns below, but did not work (nor did "spelling it out" col-span-1 ... col-span-16)

```
      pattern: /col-span-./,
      pattern: /col-span-1./,
```

**Reply by wongjn** [reply, marked as answer]
Some things to note:

You are overriding the previous `pattern` key in the same object. So:

```js
{
  …
  pattern: /col-span-14/,
  pattern: /col-span-15/,
  pattern: /col-span-16/,
}
```
is effectively:

```js
{
  pattern: /col-span-16/,
}
```
Instead, actually use regex like:

```js
{ pattern: /^col-span([1-9]|1[0-6])$/ }
```

Secondly, as per [the documentation](https://tailwindcss.com/docs/grid-column), `col-span-*` classes only go up to `12`, so you'd probably want to extend the values to go up to `16`.

**Reply by wongjn** [reply]
> The single regex seems to work -- at least up to col-span-8 (I cannot easily test more than that at the moment).
> 
> <img alt="image" width="956" src="https://user-images.githubusercontent.com/32287/256998099-80fb6be7-c550-4565-8a9d-5ac4e6ab065b.png">
> I understand your point about multiple patterns, @wongjn . But perhaps some hacky experience is interesting:
> 
> I do get a warning about col-span-16 specifically (but not warning about lower, such as col-span-15). If I keep col-span-1 ... col-span-15 and remove col-span-16, none of the large spans such as col-span-8 work (do not work). Removing one by one and testing, none of the large col-spans work until I have only col-span-1 ... col-span-8 (and no more, not even col-span-9).
> 
> <img alt="2 safelist-8-works" width="661" src="https://user-images.githubusercontent.com/32287/256997453-cc82cea8-4fb4-4412-a7cd-46dcbba837f9.png">[^1]

> I do get a warning about col-span-16 specifically […]

What is the warning? Consider checking you have implemented the solution provided correctly. Secondly, as per [the documentation](https://tailwindcss.com/docs/grid-column), `col-span-*` classes only go up to `12`, so you'd probably want to extend the values to go up to `16`.


[^1]:https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583887

**Reply by alexgenaud** [reply]
Thank you again @wongjn . I can indeed get up to col-span-12 working as expected with the regex in the safelist and without my hack.

I was trying to get classes up to col-span-16 working, but that appears to be a different issue.

<img width="846" alt="3 twelve-cols" src="https://github.com/tailwindlabs/tailwindcss/assets/32287/6aa137ca-4f22-427e-b4e4-ea5a2e0ec3ff">

**Reply by wongjn** [reply]
As mentioned in https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583782 and https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583899:

> […] as per [the documentation](https://tailwindcss.com/docs/grid-column), `col-span-*` classes only go up to `12`, so you'd probably want to extend the values to go up to `16`.

**Reply by alexgenaud** [reply]
> What is the warning?

warn - The safelist pattern `/^col-span-16$/` doesn't match any Tailwind CSS classes.

You are correct, I will not get this warning up to col-span-12, but will get this warning if I attempt to declare col-span-13 and above.

**Reply by wongjn** [reply]
As mentioned in https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583782, https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583899 and https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6584030:

> […] as per [the documentation](https://tailwindcss.com/docs/grid-column), `col-span-*` classes only go up to `12`, so you'd probably want to extend the values to go up to `16`.

**Comment by alexgenaud** [comment]
The single regex seems to work -- at least up to col-span-8 (I cannot easily test more than that at the moment).

<img width="956" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/32287/80fb6be7-c550-4565-8a9d-5ac4e6ab065b">

I understand your point about multiple patterns, @wongjn . But perhaps some hacky experience is interesting:

I do get a warning about col-span-16 specifically (but not warning about lower, such as col-span-15). If I keep col-span-1 ... col-span-15 and remove col-span-16, none of the large spans such as col-span-8 work (do not work). Removing one by one and testing, none of the large col-spans work until I have only col-span-1 ... col-span-8 (and no more, not even col-span-9).

<img width="661" alt="2 safelist-8-works" src="https://github.com/tailwindlabs/tailwindcss/assets/32287/cc82cea8-4fb4-4412-a7cd-46dcbba837f9">

**Reply by wongjn** [reply]
Replied at https://github.com/tailwindlabs/tailwindcss/discussions/11715#discussioncomment-6583899 to keep the reply chain together.

**Comment by alexgenaud** [comment]
Single regex pattern solves the issue up to col-span-12 and theme/extend/gridColumn up to col-span-16:

```
// tailwind.config.js
/** @type {import('tailwindcss').Config} */
export default {
  content: ["./src/**/*.{html,js,jsx,tsx,ts}"],
  safelist: [
    {
      pattern: /^col-span-([1-9]|1[0-6])$/
    },
  ],
  theme: {
    extend: {
      gridColumn: {
        'span-13': 'span 13 / span 13',
        'span-14': 'span 14 / span 14',
        'span-15': 'span 15 / span 15',
        'span-16': 'span 16 / span 16',
      },
    },
  },
  plugins: [],
}
```
